worker/pubsub/remoteserver.go

```diff
@@ -0,0 +1,322 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package pubsub
+
+import (
+	"sync"
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/juju/pubsub"
+	"github.com/juju/retry"
+	"github.com/juju/utils/clock"
+	"github.com/juju/utils/deque"
+	"launchpad.net/tomb"
+
+	"github.com/juju/juju/api"
+	pubsubapi "github.com/juju/juju/api/pubsub"
+	"github.com/juju/juju/apiserver/params"
+	"github.com/juju/juju/pubsub/forwarder"
+	"github.com/juju/juju/worker"
+)
+
+// RemoteServer represents the public interace of the worker
+// responsible for forwarding messages to a single other API server.
+type RemoteServer interface {
+	worker.Worker
+	UpdateAddresses(addresses []string)
+	Publish(message *params.PubSubMessage)
+}
+
+// remoteServer is responsible for taking messages and sending them to the
+// pubsub endpoint on the remote server. If the connection is dropped, the
+// remote server will try to reconnect. Messages are not sent until the
+// connection either succeeds the first time, or fails to connect. Once there
+// is a failure, incoming messages are dropped until reconnection is complete,
```

Comment:
The alternative is to guarantee delivery, which causes problems both on the sender and the receiver. The sender because you need to persist the messages somewhere, and the receiver because of dealing with old, out of date messages.

By saying that messages are dropped for non-connected agents means that means that we avoid that problem, and it just means that users of the messages need to know this, and code with that in mind.